[FEATURE] Add hex input to color picker DP-416

elskwid opened the pull request:
# Overview

Adds the ability to input the color hex value in the color picker.

Note: this modifies the behavior of the expansion panel to only toggle when the color square is clicked. This avoids some annoying behavior when clicking into the input field.

<img width="264" alt="2023-09-17-14-09-yta5a" src="https://github.com/maplight/Digital-Petitions/assets/9002/12e84fb3-8cee-4665-9577-c08be4467abd">

Fixed! ~~Note: there's probably some refinement that could happen around the click event that expands the color picker canvas, but this seems to be okay to me, for now.~~

# Ticket

[DP-416](https://maplight.atlassian.net/browse/DP-416)

[DP-416]: https://maplight.atlassian.net/browse/DP-416?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

elskwid committed:
Add direct color input to color picker

Adds the ability to directly enter the color in the color picker
component by converting the previous read-only display to an input. The
`color` property is bound to the input and a new function is added to
allow the input to emit the color so the other components can respond
accordingly.

The input itself does not do any validation. The input is validated on
save in the higher level component where these are used.

elskwid commented:
@lunox94 As you know, I don't have much experience with Angular, but I _think_ this is an okay way to make this change. I was attempting to make as few changes as possible to get the color directly editable.

drmollineda commented on src/app/features/city-staff-site-design/color-picker/color-picker.component.ts:
```diff
@@ -22,6 +22,14 @@ export class ColorPickerComponent implements OnInit {
   @Output() eventColor: EventEmitter<string | null> = new EventEmitter();
+
+  // Emit color from input
+  sendInputColor(event: Event | null) {
+    const target = event?.target as HTMLInputElement;
+    this.eventColor.emit(target.value);
+  }
```
Very minor thing, but since Angular's binding model already has the value of the input bound to `color`, you do not need to actually inspect the event for the value. You could omit the parameter in the event handler altogether and simply do the following:
```typescript
sendInputColor() {
    this.eventColor.emit(this.color);
}
```

drmollineda replied:
There is also the issue that, if edited directly, the color might not be a valid hex color string nor update the selected color in the palette. A pattern type validator can probably solve the first part.
As a future improvement, @lunox94 @elskwid, I'd recommend either moving the input field into `dp-color-palette` or pass a `FormControl` instance to it, so that validation and values can come from a single source.

elskwid replied:
@drmollineda I'll make the change to use `this.color`. Good catch. I got pretty focused on making it work, not making it right. 😁 

I don't have a strong opinion about where this should live (here or down into the `dp-color-palette`) except to say that if felt lighter to add it at the level where the color was being displayed. In terms of validation, we could definitely improve it. For now, it feels good enough that the save action checks if this is a valid hex color and reports the error to the user. What do you think?

drmollineda replied:
I think that as long as we are aware of the caveats, this one is good enough for now. I expect users might use the palette selector more often than not, unless they have a specific color string they want to use, anyway.

elskwid committed:
Remove unneeded event.target use from color-picker

Removes the retrieval of the `event.target` from the `sendInputColor`
function. The value for `color` is already bound to the model which
allows use of the color property rather than root around the event.

elskwid committed:
Update color-picker expansion panel behavior

Updates the expansion panel for the color-picker to behave a little
better in the presence of the hex color input field:

  * Expansion panel is disabled by default - this allows the user to
  click into the input without expanding the panel and prevents the
  annoying behavior of the panel toggling when if the user clicks
  in the input while the panel is expanded.
  * Adds a click event to the color display - this allows the user to
  expand the color picker/hue slider panel when desired.

elskwid commented:
@drmollineda I think this is ready for another look. Thanks! I added a new commit to deal with the expansion panel behavior weirdness.

lunox94 commented:
Neat and simple solution. I would have gone with a ReactiveForm control and its `valueChanges$` property, probably making this more complex than it really is. It's really amazing how easy is to forget the simplest and basic concepts 😅.
